proxy/envoy/ingress.go

```diff
@@ -15,256 +15,222 @@
 	"github.com/hashicorp/go-multierror"
 
-	"istio.io/api/proxy/v1/config"
+	proxyconfig "istio.io/api/proxy/v1/config"
 	"istio.io/manager/model"
 	"istio.io/manager/proxy"
 )
 
+const (
+	ingressNode = "ingress"
+	certFile    = "/etc/tls.crt"
+	keyFile     = "/etc/tls.key"
+)
+
 type ingressWatcher struct {
 	agent   proxy.Agent
-	ctl     model.Controller
-	context *IngressConfig
+	secrets model.SecretRegistry
+	mesh    *proxyconfig.ProxyMeshConfig
 }
 
 // NewIngressWatcher creates a new ingress watcher instance with an agent
-func NewIngressWatcher(ctl model.Controller, context *IngressConfig) (Watcher, error) {
-	agent := proxy.NewAgent(runEnvoy(context.Mesh, "ingress"), cleanupEnvoy(context.Mesh), 10, 100*time.Millisecond)
-
+func NewIngressWatcher(mesh *proxyconfig.ProxyMeshConfig, secrets model.SecretRegistry) (Watcher, error) {
+	agent := proxy.NewAgent(runEnvoy(mesh, ingressNode), cleanupEnvoy(mesh), 10, 100*time.Millisecond)
 	out := &ingressWatcher{
 		agent:   agent,
-		ctl:     ctl,
-		context: context,
+		secrets: secrets,
+		mesh:    mesh,
 	}
+	return out, nil
+}
 
-	if err := ctl.AppendConfigHandler(model.IngressRule, func(model.Key, proto.Message, model.Event) {
-		out.reload()
-	}); err != nil {
-		return nil, err
-	}
+func (w *ingressWatcher) Run(stop <-chan struct{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	go w.agent.Run(stop)
+	go func() {
+		<-stop
+		glog.V(2).Info("Ingress watcher terminating...")
+		cancel()
+	}()
+
+	client := &http.Client{Timeout: convertDuration(w.mesh.ConnectTimeout)}
+	url := fmt.Sprintf("http://%s/v1alpha/secret/%s/%s",
+		w.mesh.DiscoveryAddress, w.mesh.IstioServiceCluster, ingressNode)
+	initial := true
+	for {
+		tls, err := fetchSecret(ctx, client, url, w.secrets)
+		if err != nil && !initial {
+			glog.Warning(err)
+		} else {
+			config := generateIngress(w.mesh, tls, certFile, keyFile)
+			w.agent.ScheduleConfigUpdate(config)
+		}
+		initial = false
 
-	// ingress rule listing depends on the service declaration being up to date
-	if err := ctl.AppendServiceHandler(func(*model.Service, model.Event) {
-		out.reload()
-	}); err != nil {
-		return nil, err
+		select {
+		case <-time.After(convertDuration(w.mesh.DiscoveryRefreshDelay)):
+			// try again
+		case <-ctx.Done():
+			return
+		}
 	}
+}
 
+// fetchSecret fetches a TLS secret from discovery and secret storage
+func fetchSecret(ctx context.Context, client *http.Client, url string,
+	secrets model.SecretRegistry) (*model.TLSSecret, error) {
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, multierror.Prefix(err, "failed to create a request to "+url)
+	}
+	resp, err := client.Do(req.WithContext(ctx))
+	if err != nil {
+		return nil, multierror.Prefix(err, "failed to fetch "+url)
+	}
+	uri, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, multierror.Prefix(err, "failed to read request body")
+	}
+	if err = resp.Body.Close(); err != nil {
```

Comment:
Doesn't resp.Body.Close() need to be invoked regardless of whether the body contents is read, i.e. before line 102? 

You can add the following after line 100:

```golang
defer res.Body.Close() // nolint: errcheck
```